### PR TITLE
Adds ParallelGroupAFQ

### PR DESCRIPTION
**Pull request by teresamg**
Creates a ParallelGroupAFQ class inheriting from GroupAFQ to allow for easier parallelization through pydra. Does not currently work; in progress.

### CODE REVIEW

File: AFQ/api/group.py
Comment (arokem) [inline AFQ/api/group.py:919]: I don't think we're using inheritance from GroupAFQ:

```suggestion
class ParallelGroupAFQ():
```

File: AFQ/api/group.py
Comment (arokem) [inline AFQ/api/group.py:980]: I'd suggest adding: 

```
from dipy.utils.optpkg import optional_package

pydra, has_pydra, _ = optional_package('pydra')
``` 

At the top of the file for the pydra import, instead of importing within the function.

File: AFQ/api/group.py
Comment (arokem) [inline AFQ/api/group.py:934]: This is awkward. Could we change this in the implementation of `ParticipantAFQ`, instead of implementing all this renaming and deleting?

**Review by arokem — CHANGES_REQUESTED** [review]
This is a really neat implementation now! My main suggestion is to change ParticipantAFQ so that the names of kwargs and attributes are well-matched, so that you can make this even neater.

**Comment by 36000** [comment]
my thoughts, I will implement these:
1. add a similar looking `export` function to wrap export like the `export_all` implemented here
2. add pydra to setup.cfg
3. add a nighly test on 2 hbn subjects using concurrent futures for pydra

File: AFQ/api/group.py
Comment (arokem) [inline AFQ/api/group.py:974]: We should catch this error: https://github.com/nipype/pydra/issues/630 here (so that our users don't see it)

File: AFQ/api/participant.py
Comment (arokem) [inline AFQ/api/participant.py:106]: @36000 : was there a reason for these substitutions? That is, that `dwi_data_file` becomes `dwi_path`, and so on? Would there be a cost associated with this change so that internal kwargs and external names are the same?

**Comment by 36000** [comment]
I made some minor changes here. I made pydra a required library for pyAFQ (its pip installable with few dependencies, this will just be easier for most users). We now try to catch that error Ariel mentioned (let me know if this doesnt work). And I added an `export` function based on the `export_all` function. Next I will add the nightly test then I think we can merge this

**Comment by arokem** [comment]
Could you please push that test to this PR? I'm not sure what's up and would like to try to debug locally.

**Comment by 36000** [comment]
Sure, I figured out this was due to me using a different pydra version (the latest, 0.23). But now I am getting picking problems. I will push some of the changes I have made, but I think maybe we will have to talk about this in person at some point. I am running into a few issues

**Comment by 36000** [comment]
@teresamg mind trying this to see if it works?

**Comment by teresamg** [comment]
Looks like it ran successfully, both locally and on Hyak!

**Comment by arokem** [comment]
Did it generate the combined tract profiles file? I also ran a test on hyak: 

```
import os.path as op

from AFQ.api.group import ParallelGroupAFQ
from AFQ.definitions.image import RoiImage
import AFQ.api.bundle_dict as abd
import AFQ.data.fetch as afd


_, bids_path = afd.fetch_hbn_preproc(
        ["NDARZT957CWG",
         "NDARZU279XR3",
         "NDARZU401RCU",
         "NDARZU822WN3",
         "NDARZV421TCZ",
         "NDARZW262ZLV",
         "NDARZX163EWC",
         ],
        path="/gscratch/scrubbed/arokem/data/")
    
my_afq = ParallelGroupAFQ(
    bids_path=bids_path,
    preproc_pipeline="qsiprep",
    parallel_params={
        "submitter_params": {
            "plugin": "slurm",
            "sbatch_args": "-J test \
                            -p ckpt \
                            --nodes=1 \
                            --cpus-per-task=8 \
                            --gpus=1 \
                            -A escience \
                            --mem=64G \
                            --time=2:00:00 \
                            -o /gscratch/scrubbed/arokem/logs/test.out \
                            -e /gscratch/scrubbed/arokem/logs/test.err \
                            --mail-user=arokem@uw.edu \
                            --mail-type=ALL"
        },
        "cache_dir": "/gscratch/scrubbed/arokem/tmp"
    }
)

my_afq.export_all()
```

which worked great at individual subject level, but I can't find the final combined tract profiles file, which I was expecting to find.

**Comment by teresamg** [comment]
It did for me... does each subject have the *_desc-profiles_dwi.csv?

File: AFQ/api/group.py
Comment (36000) [inline AFQ/api/group.py:1120]: I am a little confused at what these lines here are trying to do

Comment (arokem) [reply AFQ/api/group.py:1120]: The idea behind this is that we want to generate a group-level csv file that contains the merged tract profiles. But we only want to do this once, so each task checks whether all of the other tasks have completed. If they have not, it returns and nothing happens. But if it is the last task running, the condition is always fulfilled and the code continues to lines 1124-1126, which run one last GroupAFQ, which would generate the combined tract profile. I worry that this is not super robust, though, and could potentially run into some funky race conditions. In my own test on hyak, I did not get the combined tract profile did not work. So, maybe we need to do something a bit more direct to create the group-level derivatives.

Comment (teresamg) [reply AFQ/api/group.py:1120]: 1120 is checking whether the process is the last process running by trawling for each subject's *_desc-profiles_dwi.csv. If all are present, it exports the GroupAFQ object to create the final tract_profiles.csv.

Comment (36000) [reply AFQ/api/group.py:1120]: OK that makes sense

Comment (36000) [reply AFQ/api/group.py:1120]: I wonder if pydra has some way to do this through their API

Comment (teresamg) [reply AFQ/api/group.py:1120]: Ariel, if all csvs are present, perhaps you can print "output_dirs" and double check that none of the paths are wonky or null? Is there anything unusual in test.out or test.err?

Comment (arokem) [reply AFQ/api/group.py:1120]: Oh yeah, good call:

```
slurmstepd: error: *** JOB 18301898 ON g3030 CANCELLED AT 2024-05-14T22:33:02 DUE TO TIME LIMIT ***
```

I will try this again with a longer max time limit than two hours.

Comment (arokem) [reply AFQ/api/group.py:1120]: I think it would be tricky to do this through pydra, because we'd have to somehow leave the submitting process running somehow for the duration of all the sub-processes, and that may become very cumbersome for very large datasets. I think that assuming we don't leave the program running until all tasks are completed, we have two options: 

1. Something like what is being done here. 
2. Don't merge and let users do that separately in a separate program. 

I suggest that if my current test works and we get what we expect that we go ahead and merge this PR as is, including this bit of code. I can follow up with a documentation example, based on my current experiments with the HBN data. Users need to know that the final merge into a tract profiles file will fail if _any_ of the sub-tasks fail, so maybe we just need to clarify this in the documentation.

Or is there anything else we need to address before merging?

Comment (36000) [reply AFQ/api/group.py:1120]: I agree. This looks good to me!

Comment (teresamg) [reply AFQ/api/group.py:1120]: I'm not sure what you mean about the submitting process... it ends as soon as the jobs are submitted, not completed, and its the last job that does the final merge (all of them check for it). Plus GroupAFQ also fails to produce tract_profiles.csv if anything fails.

File: AFQ/api/group.py
Comment (arokem) [inline AFQ/api/group.py:164]: I am getting some strange behaviors on hyak and wondering whether this is causing some issues, so I am going to remove it for now:
```suggestion
```

**Comment by arokem** [comment]
I don't know if this is directly related to the content of this PR, but I am now getting an error in visualizing the standard set of bundles, where the visualization code is raising: 

```
INFO:AFQ:Generating colorful lines from tractography...
KeyError: 'Forceps Minor'

```

Presumably because it's looking for a tract that is now no longer part of the default set of tracts (because we're using the more granular set of CC tracts). 

@36000 : any chance this is related to changes you introduced here to how data is passed between GroupAFQ and ParticipantAFQ? 

If you think this is unrelated, I think we can probably merge this PR, and fix this issue elsewhere.

**Comment by 36000** [comment]
Not sure what would cause this, but I think it is unrelated, as removing overlapping bundle definitions happens in the __init__ method of the BundleDict, so I don't think a race condition is causing this error.

**Pull request merged**
